clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionState.java

```diff
@@ -0,0 +1,159 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more contributor license agreements. See the NOTICE
+ * file distributed with this work for additional information regarding copyright ownership. The ASF licenses this file
+ * to You under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the
+ * License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+package org.apache.kafka.clients.producer.internals;
+
+import org.apache.kafka.common.TopicPartition;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.Condition;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+import static org.apache.kafka.common.requests.InitPidResponse.INVALID_PID;
+
+/**
+ * A class which maintains state for transactions. Also keeps the state necessary to ensure idempotent production.
+ */
+public class TransactionState {
+    private final PidAndEpoch pidAndEpoch;
+    private final boolean idempotenceEnabled;
+    private final Map<TopicPartition, Integer> sequenceNumbers;
+    private final Lock pidLock;
+    private final Condition hasPidCondition;
+
+    public static class PidAndEpoch {
+        public volatile long pid;
+        public volatile short epoch;
+
+        PidAndEpoch(long pid, short epoch) {
+            this.pid = pid;
+            this.epoch = epoch;
+        }
+
+        public boolean isValid() {
+            return pid != INVALID_PID;
+        }
+    }
+
+    public TransactionState(boolean idempotenceEnabled) {
+        pidAndEpoch = new PidAndEpoch(INVALID_PID, (short) 0);
+        sequenceNumbers = new HashMap<>();
+        this.idempotenceEnabled = idempotenceEnabled;
+
+        this.pidLock = new ReentrantLock();
+        this.hasPidCondition = pidLock.newCondition();
+    }
+
+    boolean hasPid() {
+        return pidAndEpoch.isValid();
```

Comment:
We can't acquire a lock here, as it would deadlock with the `pidAndEpoch` method, which acquires the lock and then calls `hasPid`. I made pidAndEpoch be volatile and immutable.